contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,284 @@
+ * * `deadline`: A timestamp after which the request is not executable anymore.
+ * * `data`: Encoded `msg.data` to send with the requested call.
+ */
+contract ERC2771Forwarder is EIP712, Nonces {
+    using ECDSA for bytes32;
+
+    struct ForwardRequestData {
+        address from;
+        address to;
+        uint256 value;
+        uint256 gas;
+        uint48 deadline;
+        bytes data;
+        bytes signature;
+    }
+
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
+        );
+
+    /**
+     * @dev Emitted when a `ForwardRequest` is executed.
+     *
+     * NOTE: A non successful forwarded request should not be assumed as non out of gas exception because of
+     * {_checkForwardedGas}. Such function doesn't guarantee an out of gas exception won't be thrown, but instead
+     * it guarantees a relayer provided enough gas to cover the signer requested gas.
+     */
+    event ExecutedForwardRequest(address indexed signer, uint256 nonce, bool success);
+
+    /**
+     * @dev The request `from` doesn't match with the recovered `signer`.
+     */
+    error ERC2771ForwarderInvalidSigner(address signer, address from);
+
+    /**
+     * @dev The requested `value` doesn't match with the available `msgValue`.
+     */
+    error ERC2771ForwarderMismatchedValue(uint256 value, uint256 msgValue);
+
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error ERC2771ForwarderExpiredRequest(uint48 deadline);
+
+    /**
+     * @dev See {EIP712-constructor}.
+     */
+    constructor(string memory name) EIP712(name, "1") {}
+
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     *
+     * A transaction is considered valid when it hasn't expired (deadline is not met), and the signer
+     * matches the `from` parameter of the signed request.
+     *
+     * NOTE: A request may return false here but it won't revert in {batchExecute} if a refund receiver
+     * is provided.
+     */
+    function verify(ForwardRequestData calldata request) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, ) = _validate(request);
+        return alive && signerMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is left stuck in the contract.
+     *
+     * Requirements:
+     *
+     * - The request value should be equal to the provided `msg.value`.
+     * - The request should be valid according to {verify}.
+     */
+    function execute(ForwardRequestData calldata request) public payable virtual returns (bool) {
+        // This check can be before the call because _execute reverts with an invalid request.
+        // Otherwise, `msg.value` will be left in the contract's balance.
```

Comment:
I find this comment confusing, I don't think it makes a difference where we put this check, does it?

Reply:
I also don't think it makes a difference but @Amxx expressed interest in doing it first:

https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4346#discussion_r1239082628

Reply:
It doesn't seem to make a difference to the behavior of the function, but we would still want the check before `_execute` to follow the principle of failing as early as possible. I just don't understand what this comment is saying so I would remove it unless we think it's saying something important then we should rephrase it.

Reply:
The comment is here because the check depends on the `_execute` having `requireValidRequest = true`, so I wanted to express that dependency that's not enforceable in code.

Reply:
Also, the check in `executeBatch` can also be done before, are we interested in consistency and following the early fail principle?

Reply:
```suggestion
        // We make sure that msg.value and request.value match exactly.
        // If the request is invalid or the call reverts, requireValidRequest ensures this value won't be lost.
```

Reply:
>the check in executeBatch can also be done before

We need to iterate through the requests array to accumulate `requestsValue` before we can do the check in `executeBatch`, so I think it makes sense where it is now.

Reply:
> We need to iterate through the requests array to accumulate requestsValue before we can do the check in executeBatch

Correct

> so I think it makes sense where it is now.

Sounds good.

I just want to mention that the for loop can be split in two without affecting gas costs (change would be marginal) and we might want to favor the early fail.